jsdoc.el

```diff
@@ -116,67 +117,96 @@
      (list
-      :name (plist-get (jsdoc--parse-param (tsc-get-child-by-field param :left)) :name)
-      :default (tsc-node-text (tsc-get-child-by-field param :right))
-      :type (jsdoc--infer-type (tsc-get-child-by-field param :right))))
-    ('array_pattern
+      :name (plist-get (jsdoc--parse-param (treesit-node-child-by-field-name param "left")) :name)
+      :default (treesit-node-text (treesit-node-child-by-field-name param "right"))
+      :type (jsdoc--infer-type (treesit-node-child-by-field-name param "right"))))
+    ("array_pattern"
      (list
       :name 'unnamed-param
-      :type (--reduce (s-concat acc "," it) (--map (plist-get (jsdoc--parse-param it) :type) (jsdoc--tsc-named-children param)))))
-    ('object_pattern
+      :type (jsdoc--infer-array-type (jsdoc--tsc-named-children param))))
+    ("object_pattern"
      (list
       :name 'unnamed-param
-      :type (--reduce (s-concat acc "," it) (--map (plist-get (jsdoc--parse-param it) :type) (jsdoc--tsc-named-children param)))))
-    ('rest_parameter
+      :type "Object"))
+    ("rest_pattern"
      (list
-      :name (tsc-node-text (tsc-get-nth-named-child param 0))
-      :type "...any"))))
+      :name (treesit-node-text (treesit-node-child param 0 t))
+      :type "...*"))))
 
 (defun jsdoc--infer-type (node)
-  (pcase (tsc-node-type node)
-    ('identifier (jsdoc--infer-identifier node))
-    ('true "boolean")
-    ('false "boolean")
-    ('number "number")
-    ('string "string")
-    ('array "any[]")
-    ('object "object")
-    ('new_expression (jsdoc--infer-type (tsc-get-nth-named-child node 0)))
-    ('call_expression (jsdoc--infer-type (tsc-get-nth-named-child node 0)))
-    ('binary_expression (jsdoc--infer-binary-expression node))
-    (_ "any")))
+  (pcase (treesit-node-type node)
+    ("identifier" (jsdoc--infer-identifier node))
+    ("true" "boolean")
+    ("false" "boolean")
+    ("number" "number")
+    ("string" "string")
+    ("array" "*[]")
+    ("object" "object")
+    ("new_expression" (jsdoc--infer-type (treesit-node-child node 0 t)))
+    ("call_expression" (jsdoc--infer-type (treesit-node-child node 0 t)))
+    ("arrow_function" (jsdoc--infer-closure-type node))
+    ("binary_expression" (jsdoc--infer-binary-expression node))
+    (_ "*")))
+
+(defun jsdoc--infer-closure-type (node)
+  "Return the inferred type of the given closure NODE."
+  (let* ((fn (jsdoc--parse-generic-function node ""))
+         (params (plist-get fn :params))
+
+         ;; Reduce the list of params if more than 1
+         (list (if (< 1 (length params))
+                   (--reduce (format "%s, %s"
+                                     (or (plist-get acc :type)
+                                         acc)
+                                     (plist-get it :type))
+                             params)
+                 (plist-get (car params) :type)
+                 ))
+         (returns (plist-get fn :returns)))
+    (concat "function(" list "): " returns)))
```

Comment:
Maybe this can be simplified a bit:

```suggestion
(defun jsdoc--infer-closure-type (node)
  "Return the inferred type of the given closure NODE."
  (let* ((fn (jsdoc--parse-generic-function node ""))
         (params (s-join ", " (--map (plist-get it :type) (plist-get fn :params))))
         (returns (plist-get fn :returns)))
    (concat "function(" params "): " returns)))
```

Reply:
Done, thanks!